src/mlpack/methods/kde/kde_impl.hpp

```diff
@@ -0,0 +1,474 @@
+
+  // Normalize if required.
+  if (kernel::KernelTraits<KernelType>::IsNormalized)
+    estimations /= kernel->Normalizer(queryTree->Dataset().n_rows);
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+void KDE<MetricType, MatType, KernelType, TreeType>::
+RelativeError(const double newError)
+{
+  CheckErrorValues(newError, absError);
+  relError = newError;
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+void KDE<MetricType, MatType, KernelType, TreeType>::
+AbsoluteError(const double newError)
+{
+  CheckErrorValues(relError, newError);
+  absError = newError;
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+template<typename Archive>
+void KDE<MetricType, MatType, KernelType, TreeType>::
+serialize(Archive& ar, const unsigned int /* version */)
+{
+  // Serialize preferences.
+  ar & BOOST_SERIALIZATION_NVP(relError);
+  ar & BOOST_SERIALIZATION_NVP(absError);
+  ar & BOOST_SERIALIZATION_NVP(breadthFirst);
+  ar & BOOST_SERIALIZATION_NVP(trained);
+
+  // If we are loading, clean up memory if necessary.
+  if (Archive::is_loading::value)
+  {
+    if (ownsKernel && kernel)
+      delete kernel;
+    if (ownsMetric && metric)
+      delete metric;
+    if (ownsReferenceTree && referenceTree)
+      delete referenceTree;
+    // After loading kernel, metric and tree, we own it.
+    ownsKernel = true;
+    ownsMetric = true;
+    ownsReferenceTree = true;
+  }
+
+  // Serialize the rest of values.
+  ar & BOOST_SERIALIZATION_NVP(kernel);
+  ar & BOOST_SERIALIZATION_NVP(metric);
+  ar & BOOST_SERIALIZATION_NVP(referenceTree);
+}
+
+template<typename MetricType,
+         typename MatType,
+         typename KernelType,
+         template<typename TreeMetricType,
+                  typename TreeStatType,
+                  typename TreeMatType> class TreeType>
+void KDE<MetricType, MatType, KernelType, TreeType>::
+CheckErrorValues(const double relError, const double absError) const
+{
+  if (relError < 0 || relError > 1)
+    throw std::invalid_argument("Relative error tolerance must be a value "
+                                "between 0 and 1");
+  if (absError < 0)
+    throw std::invalid_argument("Absolute error tolerance must be a value "
+                                "greater or equal to 0");
+  if (relError > 0 && absError > 0)
+    Log::Warn << "Absolute and relative error tolerances will be sumed up"
+              << std::endl;
```

Comment:
I don't think it's necessary to give a warning here if the documentation is clear about what happens in this case.